### PR TITLE
Add ObjectLiteralConfiguration with color/image options

### LINKED ISSUE #1587: Split rule object_literal into image_literal and color_literal
I just implemented a custom rule to prevent that any of my coworkers use the `UIImage(named: "...")` initializer since there's the `#imageLiteral` now. It looks like this:

```yaml
dynamic_image_reference:
    included: ".*.swift"
    regex: 'UIImage\(\s*named\s*:\s*\"'
    name: "Dynamic Image Reference"
    message: "Don't use dynamic image references via strings – use Xcode image literals instead."
    severity: warning
```

Then I naturally thought "why is this not part of SwiftLint anyways?" and looked it up to find the object_literal rule added via #1067. I also read the discussion there which was about adding the rule by default or not. The consensus was that `#imageLiteral` could be turned on by default but there's problems with `#colorLiteral` since some people don't use it – like we do, too. We use SwiftGen instead, which we find better since we can define the set of colors to use in a single file.

My question now is: Why not split the object_literal rule into two, namely color_literal and image_literal and turn on image_literal by default?

### PR DESCRIPTION
**Pull request by Dschee**
This fixes #1587 by implementing the discussed option configuration.

### CODE REVIEW

**Comment by Dschee** [comment]
No idea what to do to prevent Circle CI from failing – Danger tells me I need to configure `DANGER_GITHUB_API_TOKEN` but I can't find any info about that in the CONTRIBUTING.md, so not sure what to do now. Where to I add that? Do I really need it? (Other PRs seem all to run fine.) And what value should it have? I'm confused ... ^^

**Comment by marcelofabri** [comment]
@Dschee that is not your fault, but shouldn't happen anyway. @jpsim, do you know why would it fail?

Note that the Travis failure is a valid one tough.

**Comment by Dschee** [comment]
Uh, yeah, the Travis failure was due to the date, which was formed in the German style by my Xcode automatically (`03.06.17`). Just fixed it and used the weird American style instead (`06/03/17` jumping from month to day and back to year ^o^). Should pass now, hopefully!

**Comment by SDGGiesbrecht** [comment]
Merhaba, @Dschee, katkınız için teşekkür ederim. (Ich kann nur wenig Türkisch – hoffentlich bedeutet das tatsächlich was ich sagen wollte...)

> [It] was due to the date [...], which was formed in German style. [...] Just fixed it and used the weird American style instead[.]

Nach mir wären es auch eigentlich die anderen Daten die nicht stimmen. 😄 Dein lustiger Kommentar hat mich gefreut.

**Comment by Dschee** [comment]
@SDGGiesbrecht Haha, ja, ich denke das passt schon so, wobei statt "katkı" (eher für Rezepte verwendet) besser "yardım" oder "ekleme" genommen werden sollte. Aber hey, sämtliche Konjugation ist richtig. ;)

Bzgl. des Datums muss ich zugeben, dass wir im Deutschen auch solche unsinnigen Dinge haben, Beispiel die Lesart von dreistelligen Zahlen wie 365 – "drei hundert", dann "fünf und", dann Sprung zurück zu "sechzig". Also ist das Deutsche auch nicht unbedingt besser ... 😅

**Comment by Dschee** [comment]
@marcelofabri Anything here that still need to be done? I fixed the one valid issue already and consider this merge-ready.

**Comment by Dschee** [comment]
@marcelofabri The requested changes were made:

- I've added a new ObjectLiteralRuleTests file (there was none before)
- I've implemented tests for three cases: only color_literal, only image_literal, both together
- Using TDD I've added usage of the options within the rule (sorry for not seeing this before)

I've run the new tests both before it was implemented (with the only color_literal / only image_literal tests failing) and after it was implemented where the tests succeeded.

I hope this is now ready! 🎉

**Comment by marcelofabri** [comment]
> I've added a new ObjectLiteralRuleTests file (there was none before)

There was a test in `RulesTests` that now need to be removed. You should also install https://github.com/krzysztofzablocki/Sourcery and run `make sourcery` to make sure that the Linux tests are updated. More info on https://github.com/realm/SwiftLint/blob/master/CONTRIBUTING.md#code-generation

**Comment by Dschee** [comment]
I've removed the code from the `RulesTests` but I actually couldn't install sourcery, the installation instruction are really poor or the method is, I don't know. It didn't work and for now I don't have the time to deal with sourcery.

Also `script/cibuild` failed with `make: *** [test] Error 65` on `	PhaseScriptExecution Embed\ Swift\ libraries\ into\ SwiftLintFramework.framework` which doesn't give a lot of information to look into, really.

Next, when trying `swift test` I got `Can't parse Package.swift manifest file because it contains invalid format. Fix Package.swift file format and try again.`. No idea, I've not even opened Package.swift...

And `make docker_test` obviously fails because I don't have it installed. 

I'm not sure if I understood "When contributing code changes, please ensure that all three supported build methods continue to work and pass tests." correctly – cause if I really had to get all of them work and understand what they all do, that would be quite a big amount of work for contributing to this project. :D

**Comment by marcelofabri** [comment]
> I've removed the code from the RulesTests but I actually couldn't install sourcery, the installation instruction are really poor or the method is, I don't know. It didn't work and for now I don't have the time to deal with sourcery.

`brew install sourcery` should install it.

> Also script/cibuild failed with make: *** [test] Error 65 on PhaseScriptExecution Embed\ Swift\ libraries\ into\ SwiftLintFramework.framework which doesn't give a lot of information to look into, really.
> Next, when trying swift test I got Can't parse Package.swift manifest file because it contains invalid format. Fix Package.swift file format and try again.. No idea, I've not even opened Package.swift...

Are you using Xcode 8.3? The error message is probably from Xcode itself, so ¯\_(ツ)_/¯ 

> And make docker_test obviously fails because I don't have it installed.

You can install it by downloading from the website.

> I'm not sure if I understood "When contributing code changes, please ensure that all three supported build methods continue to work and pass tests." correctly – cause if I really had to get all of them work and understand what they all do, that would be quite a big amount of work for contributing to this project. Is there no simpler way?

Testing everything locally is the perfect scenario, since it avoids surprises on CI and also helps you to understand better the issue and debug it if needed. However, these checks all run on CI anyway. So unless there's something wrong, usually just running from Xcode is enough.

**Comment by Dschee** [comment]
Thanks for the help, they should really document that it's installable via Homebrew. But now that I run `make sourcery` I just get `make: `sourcery' is up to date.`. I got it so far earlier, too. But I thought that it was an issue, since I expected to see some warnings/errors or code changes happen. Nothing like that though.

Yeah, you got me, I had Xcode 9.0 beta configured on the command line. :-D Sorry for that, switching back and forth between the two recently to try out new stuff. ^^

I'm gonna wait for the CI to check the docker part, I doubt that my changes have an influence on that, but we'll see ... I can install it later if I need to. Thanks for explaining all to me, makes all sense with the CI in there. This is just not documented like this in the CONTRIBUTING.md. ;)

**Comment by Dschee** [comment]
Ok, docker actually failed, probably due to the `make sourcery` issue, I guess?

**Comment by marcelofabri** [comment]
> I'm gonna wait for the CI to check the docker part, I doubt that my changes have an influence on that

Actually, that's the main issue: you're changes are not being tested on Linux because [LinuxMain.swift](https://github.com/realm/SwiftLint/blob/master/Tests/LinuxMain.swift) wasn't updated.

> Ok, docker actually failed, probably due to the make sourcery issue, I guess?

Yes, because the list of tests from `RulesTests` wasn't updated after you removed that test.

**Comment by Dschee** [comment]
Yess, finally I got it to work! I still have no idea why `make sourcery` doesn't work for me. But running only `make` also ran the sourcery step and everything updated just fine. Now it should pass, I hope!

File: CHANGELOG.md [hunk 1]
```diff
@@ -39,6 +39,12 @@
   [Marcelo Fabri](https://github.com/marcelofabri)
   [#1546](https://github.com/realm/SwiftLint/issues/1546)
 
+* Add `code_literal` and `image_literal` boolean configuration options to
+  `object_literal` rule. They allow to only check for one or the other
+  literal type instead of both together.
```
Comment (marcelofabri) [inline CHANGELOG.md]: requires two trailing spaces as described in CONTRIBUTING.md: https://github.com/realm/SwiftLint/blob/master/CONTRIBUTING.md#tracking-changes

File: Tests/SwiftLintFrameworkTests/ObjectLiteralRuleTests.swift [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    private var allTriggeringExamples: [String] {
+        return imageLiteralTriggeringExamples + colorLiteralTriggeringExamples
+    }
+
+    // MARK: - Test Methods
+    func testObjectLiteralWithDefaultConfiguration() {
+        verifyRule(ObjectLiteralRule.description)
+    }
+
+    // swiftlint:disable:next function_body_length
```
Comment (marcelofabri) [inline Tests/SwiftLintFrameworkTests/ObjectLiteralRuleTests.swift]: I don't think you need any of these `disable:next`

File: Tests/SwiftLintFrameworkTests/ObjectLiteralRuleTests.swift [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        let baseDescription = ObjectLiteralRule.description
+        let nonTriggeringColorLiteralExamples = colorLiteralTriggeringExamples.map { example in
+            example.replacingOccurrences(of: "↓", with: "")
+        }
+
+        let description = RuleDescription(
+            identifier: baseDescription.identifier,
+            name: baseDescription.name,
+            description: baseDescription.description,
+            nonTriggeringExamples: baseDescription.nonTriggeringExamples + nonTriggeringColorLiteralExamples,
```
Comment (marcelofabri) [inline Tests/SwiftLintFrameworkTests/ObjectLiteralRuleTests.swift]: `baseDescription.nonTriggeringExamples` already contains examples similar to `nonTriggeringColorLiteralExamples`

File: CHANGELOG.md (hunk 1 above)
Comment (Dschee) [inline CHANGELOG.md]: Fixed.

File: Tests/SwiftLintFrameworkTests/ObjectLiteralRuleTests.swift (hunk 2 above)
Comment (Dschee) [inline Tests/SwiftLintFrameworkTests/ObjectLiteralRuleTests.swift]: Fixed.

**Comment by Dschee** [comment]
@marcelofabri Fixed both of your review comments.

**Comment by marcelofabri** [comment]
Thanks! 💯

**Pull request merged**